packages/protocol/lib/compatibility/verify-bytecode-foundry.ts

```diff
@@ -0,0 +1,468 @@
+ * This module verifies that deployed bytecode matches compiled Foundry artifacts.
+ * Key differences from Truffle version:
+ * - Uses Foundry artifacts with deployedBytecode.object structure
+ * - Handles Foundry's hashed library placeholders: __$<hash>$__
+ * - Computes library hashes based on source paths
+ */
+
+import { ensureLeading0x } from '@celo/base/lib/address'
+import {
+  computeFoundryLibraryHash,
+  LibraryAddressesFoundry,
+  LibraryLinksFoundry,
+  LibraryPositionsFoundry,
+  LibrarySourcePaths,
+  linkLibrariesFoundry,
+  stripMetadata,
+  verifyAndStripLibraryPrefix,
+} from '@celo/protocol/lib/bytecode-foundry'
+import { FoundryBuildArtifacts } from '@celo/protocol/lib/foundry-artifacts'
+import { verifyProxyStorageProof } from '@celo/protocol/lib/proxy-utils'
+import { ProposalTx } from '@celo/protocol/scripts/truffle/make-release'
+import Web3 from 'web3'
+import { Contract } from 'web3-eth-contract'
+import { ignoredContractsV9, ignoredContractsV9Only } from './ignored-contracts-v9'
+
+let ignoredContracts = [
+  // This contract is not proxied
+  'TransferWhitelist',
+
+  // These contracts are not in the Registry (before release 1)
+  'ReserveSpenderMultiSig',
+  'GovernanceApproverMultiSig',
+
+  // These contracts live in monorepo but are not part of the core protocol
+  'CeloFeeCurrencyAdapterOwnable',
+  'FeeCurrencyAdapter',
+  'FeeCurrencyAdapterOwnable',
+]
+
+interface VerificationContext {
+  artifacts: FoundryBuildArtifacts[]
+  libraryAddresses: LibraryAddressesFoundry
+  registry: Contract
+  governanceAddress: string
+  proposal: ProposalTx[]
+  proxyABI: any[]
+  web3: Web3
+  network: string
+  debug: boolean
+}
+
+interface InitializationData {
+  [contractName: string]: any[]
+}
+
+const ContractNameExtractorRegex = new RegExp(/(.*)Proxy/)
+const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000'
+
+// Checks if the given transaction is a repointing of the Proxy for the given contract
+const isProxyRepointTransaction = (tx: ProposalTx) =>
+  tx.contract.endsWith('Proxy') &&
+  (tx.function === '_setImplementation' || tx.function === '_setAndInitializeImplementation')
+
+export const isProxyRepointAndInitializeTransaction = (tx: ProposalTx) =>
+  tx.contract.endsWith('Proxy') && tx.function === '_setAndInitializeImplementation'
+
+export const isProxyRepointAndInitForIdTransaction = (tx: ProposalTx, registryId: string) =>
+  tx.contract === registryId && isProxyRepointAndInitializeTransaction(tx)
+
+const isProxyRepointForIdTransaction = (tx: ProposalTx, contract: string) =>
+  tx.contract === `${contract}Proxy` && isProxyRepointTransaction(tx)
+
+const isImplementationChanged = (contract: string, proposal: ProposalTx[]): boolean =>
+  proposal.some((tx: ProposalTx) => isProxyRepointForIdTransaction(tx, contract))
+
+const getProposedImplementationAddress = (contract: string, proposal: ProposalTx[]) =>
+  proposal.find((tx: ProposalTx) => isProxyRepointForIdTransaction(tx, contract)).args[0]
```

Comment:
Good idea might be to ask AI to simplify & extract / isolate somewhere such expressions into utilities to hide bloat... or maybe to avoid extracting them into functions used once 🙃